src/chainstate/stacks/boot/pox-2.clar

```diff
@@ -0,0 +1,792 @@
+      (asserts! (is-none (get-check-delegation tx-sender))
+        (err ERR_STACKING_ALREADY_DELEGATED))
+
+      ;; check valid lock period
+      (asserts! (check-pox-lock-period lock-period)
+        (err ERR_STACKING_INVALID_LOCK_PERIOD))
+
+      ;; register the PoX address with the amount stacked
+      ;;   for the new cycles
+      (try! (add-pox-addr-to-reward-cycles pox-addr first-extend-cycle extend-count amount-ustx))
+
+      ;; update stacker record
+      (map-set stacking-state
+        { stacker: tx-sender }
+        { amount-ustx: amount-ustx,
+          pox-addr: pox-addr,
+          first-reward-cycle: first-extend-cycle,
+          lock-period: lock-period })
+
+      ;; return lock-up information
+      (ok { stacker: tx-sender, unlock-burn-height: new-unlock-ht })))
+
+(define-public (delegate-stack-extend
+                    (stacker principal)
+                    (pox-addr { version: (buff 1), hashbytes: (buff 20) })
+                    (extend-count uint))
+    (let ((stacker-info (stx-account stacker))
+         (amount-ustx (get locked stacker-info))
+         (unlock-height (get unlock-height stacker-info))
+         (first-extend-cycle (burn-height-to-reward-cycle unlock-height))
+         (last-extend-cycle  (- (+ first-extend-cycle extend-count) u1))
+         (cur-cycle (current-pox-reward-cycle))
+         (lock-period (- last-extend-cycle cur-cycle))
+         (new-unlock-ht (reward-cycle-to-burn-height (+ u1 last-extend-cycle))))
+
+      ;; must be called directly by the tx-sender or by an allowed contract-caller
+      (asserts! (check-caller-allowed)
+        (err ERR_STACKING_PERMISSION_DENIED))
+
+      ;; check valid lock period
+      (asserts! (check-pox-lock-period lock-period)
+        (err ERR_STACKING_INVALID_LOCK_PERIOD))
+
+      ;; stacker must be currently locked
+      (asserts! (> amount-ustx u0)
+        (err ERR_STACK_EXTEND_NOT_LOCKED))
+
+      ;; stacker must have delegated to the caller
+      (let ((delegation-info (unwrap! (get-check-delegation stacker) (err ERR_STACKING_PERMISSION_DENIED))))
+        ;; must have delegated to tx-sender
+        (asserts! (is-eq (get delegated-to delegation-info) tx-sender)
+                  (err ERR_STACKING_PERMISSION_DENIED))
+        ;; must have delegated enough stx
+        (asserts! (>= (get amount-ustx delegation-info) amount-ustx)
+                  (err ERR_DELEGATION_TOO_MUCH_LOCKED))
+        ;; if pox-addr is set, must be equal to pox-addr
+        (asserts! (match (get pox-addr delegation-info)
+                         specified-pox-addr (is-eq pox-addr specified-pox-addr)
+                         true)
+                  (err ERR_DELEGATION_POX_ADDR_REQUIRED))
+        ;; delegation must not expire before lock period
+        (asserts! (match (get until-burn-ht delegation-info)
+                         until-burn-ht (>= until-burn-ht
+                                           new-unlock-ht)
+                      true)
```

Comment:
Why would this be an error condition?  Couldn't I revoke the delegate and re-authorize the delegate in the mean time?

Reply:
Same as above -- if your delegation has expired with the delegator, then they shouldn't be allowed to extend.

Reply:
(And to your question about `revoke -> reauthorize -> delegate-stack-extend`, yes, that would be allowed, the delegation info here is checked at the time of the `delegate-stack-extend` call, not the original `delegate-stack-stx` call).